### ISSUE TITLE
Questions about the use of mask in PGD attack

### ISSUE THREAD

**Issue by Leirunlin** [sigeisler/robustness_of_gnns_at_scale#6]
Hi! 
Thanks for your excellent work! I'm now studying your algorithm and I'm a little confused by the setting(or maybe just an implementation difference). 
In line 75 in [pgd.py](https://github.com/sigeisler/robustness_of_gnns_at_scale/blob/3c51696eb13b5440cc043949fa2fd10fd34c873d/rgnn_at_scale/attacks/pgd.py#L75), the mask seems to be the index to be attacked(e.g. the test mask). While in the [DeepRobust implementation](https://github.com/DSE-MSU/DeepRobust/blob/188caa1e4b02d4c2303a6f1f3e7c2cd5bca234cb/deeprobust/graph/global_attack/topology_attack.py#L114), the training mask is used. 
Does the choice of mask matter here? 
By the way, I noticed that the validation mask seems to be ignored by attack algorithms while being used by the defense algorithm. Is this also a common setup?

Best, 
Runlin.

**Comment by sigeisler** [comment]
Hi @Leirunlin,

Thank you for reaching out! 

I am not sure why they call it `idx_train`, as in the end, it is up to the user what indices to pass to the `attack` method. For this reason, we call it `idx_attack` instead.

I think their mindset is that you run an evasion attack (training time) on the train labels and then use this for training a new model on the "poisoned" graph.

Anyhow, if you look into their [example](https://github.com/DSE-MSU/DeepRobust/blob/188caa1e4b02d4c2303a6f1f3e7c2cd5bca234cb/examples/graph/test_pgd.py#L95), they concatenate training and test to use it as `idx_train`. Since `idx_train` is much smaller on the common datasets than `idx_test`, it is probably about the same as solely using `idx_test`.

In summary, it really depends on what you want to achieve to choose what `idx_attack` should mean.

Please let me know if you have further questions.

PS: if you have poisoning in mind also check out the example in my current [PyG pull request (hopefully it will be released with the next version)](https://github.com/pyg-team/pytorch_geometric/pull/5972).

**Comment by Leirunlin** [comment]
Thanks for your detailed reply! Most of my questions have been well addressed.
It seems that I've been confused between "poison attacks" and "transferring from an evasion attack to the poison setting". As an example of the former case, the [MinMax attack](https://github.com/DSE-MSU/DeepRobust/blob/188caa1e4b02d4c2303a6f1f3e7c2cd5bca234cb/examples/graph/test_min_max.py)(a poisoned version of PGD) only uses `idx_train`. The PGD algorithm we discussed is more about the latter case, and more choices of masks could be considered. (Hope I don't misunderstand it again)
And the usage of the validation set is still unclear to me. It would be so nice of you if any suggestions could be possible. 
At last, the PyG example looks really great! I will check out the examples. Wish them to be released soon.

Best,
Runlin.

**Comment by sigeisler** [comment]
Hi @Leirunlin,

I think you can ignore the validation set unless you want to attack e.g. the early stopping mechanism / hyperparameter selection of your training etc. These should be the only occasions when the validation performance matters.

I think the [PyG poisoning example](https://github.com/pyg-team/pytorch_geometric/pull/5972/files#diff-481cf8987553762c276802e9f867632fb4e097872b511d7d8202322b5b2f1021) is more promising than the MinMax attack (at least in my experience). You can always install PyG from git (i.e. `pip install git+https://github.com/sigeisler/pytorch_geometric.git@ac43df93966c05d8422d8e0ea78c30bfea7e2842`)
